### PR TITLE
CORDA-2688 - Add Serialization Context option for no carpenting

### PR DESCRIPTION
**Pull request by fenryka**
Can be used by the attachment class loader - Serialisation Framework will still consume all Exceptions and throw a NotSerializableException

@tudor-malene unsure how to test this from your perspective, but the carpentry unit tests now try
without it enabled and assert it throws a NotSerializableException

### CODE REVIEW

File: serialization/src/main/kotlin/net/corda/serialization/internal/amqp/ObjectSerializer.kt [hunk 1]
```diff
@@ -17,7 +17,7 @@ interface ObjectSerializer : AMQPSerializer<Any> {
             if (typeInformation is LocalTypeInformation.NonComposable)
                 throw NotSerializableException(
                         "Trying to build an object serializer for ${typeInformation.typeIdentifier.prettyPrint(false)}, " +
-                        "but it is not constructible from its public properties, and so requires a custom serialiser.")
+                        "but it is not constructable from its public properties, and so requires a custom serialiser.")
```
Comment (r3domfox) [inline serialization/src/main/kotlin/net/corda/serialization/internal/amqp/ObjectSerializer.kt]: https://www.google.com/search?q=constructable

File: serialization/src/main/kotlin/net/corda/serialization/internal/model/TypeLoader.kt [hunk 2]
```diff
@@ -25,14 +26,20 @@ class ClassCarpentingTypeLoader(private val carpenter: RemoteTypeCarpenter, priv
...
+    ): Map<TypeIdentifier, Type> {
         val remoteInformationByIdentifier = remoteTypeInformation.associateBy { it.typeIdentifier }
 
         // Grab all the types we can from the cache, or the classloader.
         val noCarpentryRequired = remoteInformationByIdentifier.asSequence().mapNotNull { (identifier, _) ->
             try {
                 identifier to cache.computeIfAbsent(identifier) { identifier.getLocalType(classLoader) }
             } catch (e: ClassNotFoundException) {
+                if (context.carpenterDisabled) {
+                    throw e
```
Comment (tudor-malene) [inline serialization/src/main/kotlin/net/corda/serialization/internal/model/TypeLoader.kt]: This is neat!
I was expecting the carpenter to be more difficult to extract.

**Comment by fenryka** [comment]
> Given that only the "allow carpentry" flag on the context is actually _used_ by the type loader, might we narrow the interface and pass in just the flag, rather than the entire context object?

We could certainly do that, but passing around the object isn't any less efficient (save 3 whole bytes) and it could be useful later to have the context object passed down the callstack... if you're super keen on not doing it I'll put it back the other way

File: serialization/src/main/kotlin/net/corda/serialization/internal/amqp/ObjectSerializer.kt (hunk 1 above)
Comment (fenryka) [reply serialization/src/main/kotlin/net/corda/serialization/internal/amqp/ObjectSerializer.kt]: Ah, but IntelliJ understands one and not the other ;)

**Comment by r3domfox** [comment]
It might make sense to refactor towards a narrower interface if it starts to be used to do more things (and thereby confusing the purpose of `SerializationContext`), but for the time being it's harmless and easy enough to change if we find we need to.

File: serialization/src/main/kotlin/net/corda/serialization/internal/model/TypeLoader.kt (hunk 2 above)
Comment (fenryka) [reply serialization/src/main/kotlin/net/corda/serialization/internal/model/TypeLoader.kt]: Once I dug into it... thankfully not

**Pull request merged**